src/Std/Data/HashMap/Lemmas.lean

```diff
@@ -751,6 +751,68 @@ theorem distinct_keys_toList [EquivBEq α] [LawfulHashable α] :
     m.toList.Pairwise (fun a b => (a.1 == b.1) = false) :=
   DHashMap.Const.distinct_keys_toList
 
+section monadic
+
+variable {m : HashMap α β} {δ : Type w} {m' : Type w → Type w}
+
+theorem foldM_eq_foldlM_toList [Monad m'] [LawfulMonad m']
+    {f : δ → (a : α) → β → m' δ} {init : δ} :
+    m.foldM f init = m.toList.foldlM (fun a b => f a b.1 b.2) init :=
+  DHashMap.Const.foldM_eq_foldlM_toList
+
+theorem fold_eq_foldl_toList {f : δ → (a : α) → β → δ} {init : δ} :
+    m.fold f init = m.toList.foldl (fun a b => f a b.1 b.2) init :=
+  DHashMap.Const.fold_eq_foldl_toList
+
+theorem foldRevM_eq_foldrM_toList [Monad m'] [LawfulMonad m']
+    {f : δ → (a : α) → β → m' δ} {init : δ} :
+    m.foldRevM f init = m.toList.foldrM (fun a b => f b a.1 a.2) init :=
+  DHashMap.Const.foldRevM_eq_foldrM_toList
+
+theorem foldRev_eq_foldr_toList {f : δ → (a : α) → β → δ} {init : δ} :
+    m.foldRev f init = m.toList.foldr (fun a b => f b a.1 a.2) init :=
+  DHashMap.Const.foldRev_eq_foldr_toList
+
+theorem forM_eq_forM_toList [Monad m'] [LawfulMonad m'] {f : (a : α) → β → m' PUnit} :
+    m.forM f = m.toList.forM (fun a => f a.1 a.2) :=
+  DHashMap.Const.forM_eq_forM_toList
+
+theorem forIn_eq_forIn_toList [Monad m'] [LawfulMonad m']
```

Comment:
Similar to the setup for `List`, the left-hand side of this lemma should be stated in terms of `ForIn.forIn`, and there should be an additional `@[simp]` lemma that rewrites `HashMap.forIn` to `ForIn.forIn`.